File: contracts/src/main/java/my/restate/e2e/services/Counter.java
// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate e2e tests,
// which are released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/e2e/blob/main/LICENSE

package my.restate.e2e.services;

import dev.restate.sdk.ObjectContext;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.VirtualObject;

@VirtualObject
public interface Counter {

  @Handler
  void add(ObjectContext context, long value);

  @Handler
  void reset(ObjectContext context);

  @Handler
  void addThenFail(ObjectContext context, long value);

  @Handler
  long get(ObjectContext context);

  @Handler
  CounterUpdateResponse getAndAdd(ObjectContext context, long value);

  @Handler
  void infiniteIncrementLoop(ObjectContext context);
}

File: sdk-tests/src/main/kotlin/dev/restate/sdktesting/contracts/Counter.kt
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate SDK Test suite tool,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/sdk-test-suite/blob/main/LICENSE
package dev.restate.sdktesting.contracts

import dev.restate.sdk.annotation.*
import kotlinx.serialization.Serializable

@VirtualObject
@Name("Counter")
interface Counter {
  @Serializable data class CounterUpdateResponse(val oldValue: Long, val newValue: Long)

  /** Add value to counter */
  @Handler suspend fun add(value: Long): CounterUpdateResponse

  /** Add value to counter, then fail with a Terminal error */
  @Handler suspend fun addThenFail(value: Long)

  /** Get count */
  @Shared suspend fun get(): Long

  /** Reset count */
  @Handler suspend fun reset()
}
